tests/test_software_upgrade.sh

```diff
@@ -0,0 +1,169 @@
+    upgrade_info="{\"binaries\":{\"linux/amd64\":\"$download_path/gaiad-$upgrade_version-linux-amd64\",\"linux/arm64\":\"$download_path/$upgrade_version/gaiad-$upgrade_version-linux-arm64\",\"darwin/amd64\":\"$download_path/gaiad-$upgrade_version-darwin-amd64\",\"windows/amd64\":\"$download_path/gaiad-$upgrade_version-windows-amd64.exe\"}}"
+    proposal="gaiad tx gov submit-proposal software-upgrade $upgrade_name --from $validator_address --keyring-backend test --upgrade-height $upgrade_height --upgrade-info $upgrade_info --title gaia-upgrade --description 'test' --chain-id my-testnet --deposit 10stake --yes"
+
+    # Submit the proposal and vote yes for it
+    echo "Submitting the upgrade proposal."
+    echo $proposal
+    $proposal
+
+    echo "Submitting the \"yes\" vote."
+    vote="gaiad tx gov vote 1 yes --from $validator_address --keyring-backend test --chain-id my-testnet --sequence $validator_sequence --yes"
+    echo $vote
+    $vote
+
+    # Wait for the voting period to be over
+    echo "Waiting for the voting period to end..."
+    sleep $voting_waiting_time
+
+    echo "Upgrade proposal status:"
+    gaiad q gov proposal 1 --output json | jq '.status'
+
+    # Wait until the right height is reached
+    height=0
+    max_tests=60
+    test_counter=0
+    echo "Waiting for the upgrade to take place at block height $upgrade_height..."
+    until [ $height -ge $upgrade_height ]
+    do
+        sleep 5
+        if [ ${test_counter} -gt ${max_tests} ]
+            then
+            echo "Testing gaia for $test_counter times but did not reach height $upgrade_height"
+            exit 2
+        fi
+        height=$(curl -s http://$gaia_host:$gaia_port/block | jq -r .result.block.header.height)
+        if [ -z "$height" ]
+        then
+            height=0
+        fi
+        echo "Block height: $height"
+        test_counter=$(($test_counter+1))
+
+    done
+    echo "The upgrade height was reached."
+
+    # Waiting until gaiad responds
+    attempt_counter=0
+    max_attempts=60
+    echo "Waiting for gaia to come back online..."
+    until $(curl --output /dev/null --silent --head --fail http://$gaia_host:$gaia_port)
+    do
+        if [ ${attempt_counter} -gt ${max_attempts} ]
+        then
+            echo ""
+            echo "Tried connecting to gaiad for $attempt_counter times. Exiting."
+            exit 3
+        fi
+
+        printf '.'
+        attempt_counter=$(($attempt_counter+1))
+        sleep 1
+    done
+
+    # Get running version
+    gaiad_upgraded_version=$(curl -s http://$gaia_host:$gaia_port/abci_info | jq -r .result.response.version)
+    echo "Current gaiad version: $gaiad_upgraded_version"
+
+    # Check upgraded version is the one we want
+    if [[ "$gaiad_upgraded_version" != "$upgrade_version" ]]; then
+        echo "Requested $upgrade_version, but detected $gaiad_upgraded_version."
+        exit 4
+    fi
+
+    # Check if gaiad is producing blocks
```

Comment:
We can merge right now, but I think in a future iteration we can re-use the test_block_production code here. There's a bit of redundancy

Reply:
Agreed, we can break up these two scripts next.